### ISSUE TITLE
VPP Dataplane: memif mechanism

### ISSUE THREAD

**Issue by edwarnicke** [ligato/networkservicemesh#341]
This issue is to integration the memif mechanism into the vpp dataplane:

https://github.com/ligato/networkservicemesh/tree/master/dataplanes/vpp

I'll add a sequence of comments here to explain things that are helpful to understand of think about to accomplish this goal.

**Comment by edwarnicke** [comment]
# Understanding memif

memif is an 'interface' for carrying L2 or L3 packets from one process running on a server to another process running on a server.  It does not show up like a kernel interface in things like ifconfig.  Instead it operates by using shared memory between both processes.

In order to accomplish this, both processes must have access to a common unix file socket.  One of those processes must be designated 'master' and the other 'slave'.  The processes use the unix file socket to negotiate setup, and for one to pass the other file descriptors (fds) for shared memory to be used between them.  The file socket is only used for setup.

Thus, in order for two processes to setup a 'memif' connection between them, they must agree on two things:

1. What unix file socket to use
2. Both of them have to be able to access the file socket
2. Which of them is master and which is slave.

# Understanding memif in containers

Containers are, in some sense, just two processes running on the same server, with some interesting segregation.  Thus they have to agree on the same three things.

**Comment by edwarnicke** [comment]
# Understanding shared directories in Network Service Mesh

Each Pod seeking to use NSM has to request a resource [nsm.ligato.io](https://github.com/ligato/networkservicemesh/blob/master/conf/sample/nsm-client.yaml#L91).

This causes kubelet to make an Allocate call using the [Device Plugin API](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/).  The Allocate call, among other things, returns a [Mount](https://github.com/ligato/networkservicemesh/blob/master/plugins/nsmserver/nsmserver_client.go#L387) which is then mounted into the containers in the Pod: they can all access it.  Each Pod gets its own directory mounted in, which is unique to that Pod as long as it is running.  These directories can and do get recycled and used for a new Pod once the one they are assigned to finishes running.  So whatever you do in that directory, make sure to clean it up before using it again in the [Allocate](https://github.com/ligato/networkservicemesh/blob/master/plugins/nsmserver/nsmserver_client.go#L371) call.

I would recommend creating a memif/ subdir in this directory that is shared between the Pod and the NSM into which you can put memif socket files.

**Comment by edwarnicke** [comment]
# Sharing the NSM directory with the Dataplane

Since the [Dataplane ](https://github.com/ligato/networkservicemesh/tree/master/dataplanes/vpp) is responsible for creating the memif/ subdirs in the per-Pod subdirs, and since those all live in the [volume](https://github.com/ligato/networkservicemesh/blob/master/conf/sample/networkservice-daemonset.yaml#L36) mounted into the NSM at [/var/lib/networkservicemesh](https://github.com/ligato/networkservicemesh/blob/master/conf/sample/networkservice-daemonset.yaml#L29) it is recommended that you have the vpp dataplane similarly mount it.

This will allow you access to the per-Pod subdirs, and allow you to create memif unix file sockets where the Pods can get at them.  This solves:

1. What unix file socket to use
2. Both of them have to be able to access the file socket

**Comment by edwarnicke** [comment]
# A word on direct cross connects

memif's have the magical ability to allow two containers to directly cross connect if they are on the same host and both using memif without needing to go through a vswitch.  This is a *huge* advantage.

In the direct cross connect, you have a challenge: a single unix file socket has to be in the memif/ directory for *two* Pods.  This should be solvable by creating it in the memif/ directory for one Pod, and then creating a hardlink to that socket in the memif/ directory for the other Pod.

**Comment by edwarnicke** [comment]
# The K8s NSM Dataplane API

The [K8s NSM Dataplane API](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/dataplaneinterface/dataplaneinterface.proto) fundamentally is about Creating and Destroying [Connections](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/dataplaneinterface/dataplaneinterface.proto#L21).  A Connection has a [local_source](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/dataplaneinterface/dataplaneinterface.proto#L22) that is a [LocalMechanism](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/common/common.proto#L31) that has a [Type](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/common/common.proto#L36) and [parameters](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/common/common.proto#L33). parameters are simply a map<string,string>.

This issue is about the [MEM_INTERFACE](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/common/common.proto#L40) type (which should be renamed MEMIF).

When a Network Service Client Pod seeks to open a connection, it will provide a preference ordered list of partially specified [LocalMechanism](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/nsmconnect/nsmconnect.proto#L15).  Being partially specified means it has some, but not necessarily all of the needed parameters filled in for the requested LocalMechanism.  When the Network Service Manager [returns the one LocalMechanism being used](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/nsmconnect/nsmconnect.proto#L30), that LocalMechanism must be fully specified.

For type MEMIF, the parameters are:

* socketfile - name of the socketfile relative to the memif/ directory for the Pod
* master - master=true indicates the LocalMechanism is memif master
* slave - slave=true indicates the LocalMechanism is memif slave

The parameters master and slave should not be both set at the same time on the same LocalMechanism.

If the VPP Dataplane receives a LocalMechanism without the socketfile specified, it may choose the name itself, subject to non-collision in the same memif directory.

If the VPP Dataplane receives a LocalMechanism without the master or slave set, it may chose that itself, subject to the constraint that one side of a memif must be master, and the other slave.

**Comment by edwarnicke** [comment]
@sbezverk VPP does not *strictly* have to be master.  We have two cases:

1.  direct memif to memif - in this case, one side is picked master the other slave
2.  memif to vpp vswitch to other mechanism - in this case, as long as one side is master and the other slave, we are fine.  Generally I'd personally tend towards making vpp master, but if the NSE or NSC had a strong opinion, I'd not argue.

**Comment by edwarnicke** [comment]
# Enhancement to ConnectionRequest Message for Pod directory

I'd suggest looking at enhancing the [ConnectionRequest message here](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/nsmconnect/nsmconnect.proto#L14) to provide the Pod directory.  I'd recommend providing it relative to the /var/lib/networkservicemesh directory being mounted in both the NSM and the Dataplane.

**Comment by sbezverk** [comment]
Not interested in direct memif because it totally breaks logic. 2 scenario allows to build reusable and flexible code. Also it allows to seamlessly apply features to any time of connectivity, as long as they follow path client1 --- vpp --- client2

**Comment by edwarnicke** [comment]
@sbezverk Direct memif is actually *crucial* for the Cloud-native NFV case... its hugely faster than going through the vswitch.  Its also the primary case being tested by the VNF/CNF comparison folks.

**Comment by sbezverk** [comment]
@edwarnicke I guess you know better, so maybe as an exception, but the general approach should be a common path for as many type of connectivity as possible.. imho of course..

**Comment by edwarnicke** [comment]
@sbezverk I get your point :)  And usually agree... but a huge selling point for us is the memif/memif direct connect :)

**Comment by edwarnicke** [comment]
# Manipulating memif via CLI

We will be programming memif using the Go language bindings, but it helps tremendously to be able to manipulate it via the CLI for purposes of debugging and trial and error.

VPP provides a command

```
vppctl ${cmd}
```

If vpp is running in a docker container:

```
docker exec ${dockerid} vppctl ${cmd}
```

The [cli commands related to memif](https://docs.fd.io/vpp/18.07.1/clicmd_src_plugins_memif.html).
Note: for our case we will have to:

```
vppctl create memif socket ...
```

before we
```
vppctl create interface memif ...
```

**Comment by edwarnicke** [comment]
# Programming VPP memif via Go

The [memif Go API is available already in the repo](https://github.com/ligato/networkservicemesh/blob/master/dataplanes/vpp/bin_api/memif/memif.ba.go)

**Comment by edwarnicke** [comment]
# Overall cross connects

Implementing the memif mechanism means implementing cross connects:

1. memif to memif
2. memif to kernel interface

(note: kernel interface to kernel interface mechanism has already been done).

**Comment by edwarnicke** [comment]
# Where to start in the VPP Dataplane

[ConnectionRequest](https://github.com/ligato/networkservicemesh/blob/master/pkg/nsm/apis/nsmconnect/nsmconnect.proto#L8) is [implemented in the VPP Dataplane](https://github.com/ligato/networkservicemesh/blob/master/dataplanes/vpp/pkg/nsmdataplane/nsmdataplane_server.go#L67).  This is probably the best place to start in the Dataplane.

**Comment by edwarnicke** [comment]
This issue relies on Issue #343 to be fully testable.

**Comment by edwarnicke** [comment]
# Understanding bind mounts

Turns out that a memif master will always delete and existing unix file socket, and create a new one.  This breaks the proposed mechanism for direct memif-memif in [A Word on Direct Cross Connects](https://github.com/ligato/networkservicemesh/issues/341#issuecomment-429681049).   

But fear not, there is another solution.  Bind mounts.

A Bind Mount allows you to 'mount' something into an additional place in the directory tree.

```bash
mkdir -p /tmp/foo
mkdir -p /tmp/bar
mount --bind /tmp/foo /tmp/bar
touch /tmp/foo/baz
ls -id /tmp/bar
ls -id /tmp/foo
```
will show you that /tmp/foo and /tmp/bar have the same inode.  They are in fact the same directory.

**Comment by edwarnicke** [comment]
# Bind mounts and direct memif memif

For direct memif memif connections if we have two Pods: Pod1, Pod2, and

Pod1 has Mount into it /var/lib/networkservicemesh/nsm-1

Pod2 has Mount into it /var/lib/networkservicemesh/nsm-2

Then the Dataplane Pod can create a directory:

```/var/lib/networkservicemesh/nsm-1/memif/1-2/```

and then bind mount it

```bash
mkdir -p /var/lib/networkservicemesh/nsm-2/memif/2-1/
mount --bind /var/lib/networkservicemesh/nsm-1/memif/1-2/ /var/lib/networkservicemesh/nsm-2/memif/2-1/
```
This will make sure that 
```/var/lib/networkservicemesh/nsm-2/memif/2-1/``` and ```/var/lib/networkservicemesh/nsm-1/memif/1-2/``` are in fact the same directory.  One of them will be visible to Pod1, the other to Pod2.

This way, if the memif master is in Pod1 it can create the memif socket file in the directory:

/var/lib/networkservicemesh/nsm-1/memif/1-2/

and it will be visible in Pod2 at:

/var/lib/networkservicemesh/nsm-2/memif/2-1/

**Comment by edwarnicke** [comment]
# Bind mount propagation and Kubernetes

There is a tricky corner with bind propagation and Kubernetes.  Normal volumes get mounted into Pods by Kubernetes via bind mounting, and so Dataplane will be creating a bind mount under a bind mount.

In order to have this result in that bind mount being visible on other Pods, you have to set:
[mountPropagation: Bidirectional](https://kubernetes.io/docs/concepts/storage/volumes/#mount-propagation) for the volumeMount.  [This](https://github.com/ligato/networkservicemesh/pull/373/commits/b1610119176d2357ac63fad52e5e76a0d9143511) shows how to do it.

[There is a really in-depth explanation of mount propagation here](https://medium.com/kokster/kubernetes-mount-propagation-5306c36a4a2d) if you are curious.

**Comment by edwarnicke** [comment]
# Golang and bind mounts

[go mount](https://godoc.org/github.com/docker/docker/pkg/mount) appears to support [BIND](https://godoc.org/github.com/docker/docker/pkg/mount#pkg-constants).

**Comment by mestery** [comment]
I think this is addressed by #360.

**Comment by edwarnicke** [comment]
Was this completed by #485 ?  If so, should we close the issue?
